src/Build/Evaluation/Evaluator.cs

```diff
@@ -1790,8 +1791,33 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
                 if (solutionPath == "*Undefined*") solutionPath = null;
                 var projectPath = _data.GetProperty(ReservedPropertyNames.projectFullPath)?.EvaluatedValue;
 
+                var compareInfo = CultureInfo.InvariantCulture.CompareInfo;
+
+                static bool HasProperty(string value, CompareInfo compareInfo) =>
+                    value != null && compareInfo.IndexOf(value, "$(") != -1;
+
+                if (HasProperty(sdkReference.Name, compareInfo) ||
+                    HasProperty(sdkReference.Version, compareInfo) ||
+                    HasProperty(sdkReference.MinimumVersion, compareInfo))
+                {
+                    static string EvaluateProperty(string value, IElementLocation location, Expander<P, I> expander) =>
+                        value != null
+                            ? expander.ExpandIntoStringAndUnescape(value,
```

Comment:
I could somewhat imagine this being problematic if you want your name to include a %, and it wasn't escaping it previously. If your name includes $(, you're probably asking for it if you don't expect it to be expanded. Is it feasible for it to be a path intended for use as a key at some point? Because ExpandIntoStringAndUnescape also fixes file paths.

Reply:
You propose using `ExpandIntoStringLeaveEscaped` instead, right? So that it would preserve special characters, like `%`, not treat them like a part of escaped character sequence. I'm not sure about this part, I leave it for MSBuild team to decide. If you think it's better to use the other option, I'll gladly make the change.

Reply:
@Forgind, trying to summarize the failure cases. Let me know if I've missed any:
- user has a `$(` which they don't want expanded, but now it will get expanded. The traditional approach here is for the user to escape any msbuild syntax when they want it as a literal, which will still work here.
- on non windows systems, the `$(` is [used as a guard to prevent adjusting path like strings](https://github.com/cdmihai/msbuild/blob/9c2d922959e80ba72fefb7addec21ede227639c3/src/Shared/FileUtilities.cs#L454). So unix based users with strings that contained both `$(` and backslashes in their sdk references, will now get their slashes changed. I think this is still fine:
  - first, `$(` and slashes would constitute illegal nuget names and versions, so their builds would have probably crashed anyway had they tried to put these in nuget based sdks.
  - second, even though backslashes are valid unix filename characters, AFAIK their usage is not common. And even if the user did have backslashes in their name / version, they can escape them. If this gets to be a common case (and thus make escaping bad for usability) we could then add an option to not fix the file path when expanding the sdk reference (my guess is it won't happen).

Reply:
These are certainly all things the user can get out of pretty easily; I'm just trying to avoid having code that works with one MSBuild version, then fails with the next. I was also thinking of the failure case @andrew-boyarshin mentioned with %. The most odd part to me is if you have, say, a name with an escape character but no properties. Then you add a property to the version, and the name is escaped. That also applies to paths, so you noted that slashes would be illegal—are quotes? Since MaybeAdjustFilePath also removes quotes if present. I think that was all I was thinking. I'm not too worried about most properties being expanded, but it did occur to me that if someone uses the guard against paths being adjusted, it might look like a property and be converted to "" when it can't be found. That's the worst case I can imagine.

Reply:
> trying to avoid having code that works with one MSBuild version, then fails with the next

If it does break anybody, it's probably in their own interest to fix it, since the correct behaviour from the very start should have been to expand the sdk reference. So they'd be migrating off of relying on a bug.

However, in case it does break anybody, it probably makes sense to enable the expansion by default but provide an escape hatch so if anybody does break, they can revert to the old behaviour. @andrew-boyarshin, can you please add an escape hatch [here](https://github.com/dotnet/msbuild/blob/f365aa332f58cf43e78d76491438f15fc1988f27/src/Shared/Traits.cs#L110), and disable sdk reference expansion if the escape hatch is enabled?

>MaybeAdjustFilePath also removes quotes if present

I don't think it does. The results from RemoveQuotes does not seem to flow into the return of MaybeAdjustFilePath

>a name with an escape character but no properties. Then you add a property to the version, and the name is escaped.

This breaking scenario [is constrained ](https://github.com/cdmihai/msbuild/blob/9c2d922959e80ba72fefb7addec21ede227639c3/src/Deprecated/Engine/Shared/EscapingUtilities.cs#L65-L66) only to strings that have `%` followed by two digits. In which case (beyond the fact nuget would probably fail on it), the user can use the escape hatch.